### PR TITLE
MODCAL-45 Split permissions by http method

### PR DESCRIPTION
**Pull request by dmytrotur**
## Purpose
Resolves: [MODCAL-45](https://issues.folio.org/browse/MODCAL-45)
Split permissions by HTTP method in module descriptor
Corresponding pull requests:

- https://github.com/folio-org/mod-circulation/pull/284
- https://github.com/folio-org/ui-calendar/pull/194

### CODE REVIEW

**Comment by dmytrotur** [comment]
@marcjohnson-kint 
> Why has the interface version number been changed? Is it to act as a marker to dependents that they need to change their permissions?

I considered it to be a breaking change, and increased interface version. But, in fact, several endpoints just require less permissions after this change.
Do you think we can leave interface version unchanged?

**Comment by dmytrotur** [comment]
> Was renaming the permissions (to better fit with FOLIO's conventions) consider for part of this change?

No, it wasn't. But I think it can be done here.  
I can only think about changing them in the following way:
`calendar.collection.get` - leave as is
`calendar.collection.add` -> `calendar.collection.post` 
`calendar.collection.update` -> `calendar.collection.put` 
`calendar.collection.remove` -> `calendar.collection.delete` 
Please, leave your suggestions if you have.

**Comment by marcjohnson-kint** [comment]
> `calendar.collection.update` -> `calendar.collection.put`

Is the put for a single entry in the calendar? If so, `calendar.item.put` might be more appropriate. This doesn't read write at the moment. What type of record do these endpoints use?

> `calendar.collection.remove` -> `calendar.collection.delete`

Does this delete a single calendar entry?

**Comment by marcjohnson-kint** [comment]
> @marcjohnson-kint
> 
> > Why has the interface version number been changed? Is it to act as a marker to dependents that they need to change their permissions?
> 
> I considered it to be a breaking change, and increased interface version. But, in fact, several endpoints just require less permissions after this change.
> Do you think we can leave interface version unchanged?

This is one of the contentious or ambiguous areas of design in FOLIO. I think it is a subtle coupling we've introduced between a dependent module and an implementation of an interface.

I think required permissions are part of the declaration that an implementation makes when it binds itself to an interface. Hence I think it is part of the implementation, and is an implementation version change rather than an interface version change.

Given that this change _weakens_ the expectations of a client, I think this is a minor change. 

That said, if we rename the permissions, that likely becomes breaking.

@skoczko What do you think?

**Comment by dmytrotur** [comment]
@marcjohnson-kint 
`calendar.collection.get` -> split to: 
`calendar.opening-hours.collection.get`
 `calendar.periods.collection.get` 
`calendar.periods.item.get`

`calendar.collection.add` -> `calendar.periods.item.post` 
`calendar.collection.update` -> `calendar.periods.item.put` 
`calendar.collection.remove` -> `calendar.periods.item.delete` 
Does this change seems to be appropriate?

**Comment by marcjohnson-kint** [comment]
> @marcjohnson-kint
> `calendar.collection.get` -> split to:
> `calendar.opening-hours.collection.get`
> `calendar.periods.collection.get`
> `calendar.periods.item.get`
> 
> `calendar.collection.add` -> `calendar.periods.item.post`
> `calendar.collection.update` -> `calendar.periods.item.put`
> `calendar.collection.remove` -> `calendar.periods.item.delete`
> Does this change seems to be appropriate?

Yes, I think those make sense. Under the new naming, I think `calendar.collection.get` would become `calendar.periods.collection.get`, is that the case?

**Comment by dmytrotur** [comment]
> Under the new naming, I think calendar.collection.get would become calendar.periods.collection.get, is that the case?

`calendar.collection.get` permission was required by several endpoints in calendar. I split it so that each endpoint has separate required permission. Talking about mod-circulation, we need to change `calendar.collection.get`  to `calendar.opening-hours.collection.get` (already done in mod-circulation pull request).

File: descriptors/ModuleDescriptor-template.json [hunk 1]
```diff
@@ -5,39 +5,60 @@
...
+      "version": "3.0",
       "handlers": [
         {
           "methods": [
             "GET"
           ],
           "pathPattern": "/calendar/periods",
           "permissionsRequired": [
-            "calendar.collection.get"
+            "calendar.opening-hours.collection.get"
```
Comment (marcjohnson-kint) [inline descriptors/ModuleDescriptor-template.json]: Why is the permission for `/calendar/periods` naming `opening-hours`, yet `/calendar/periods/{servicePointId}/period` is named `periods`?

What type of record comes from each of these two endpoints?

Comment (dmytrotur) [reply descriptors/ModuleDescriptor-template.json]: I'm not sure about the reasons why this naming was used for these endpoints, but actually `/calendar/periods` returns `opening-hour` records, 
when `/calendar/periods/{servicePointId}/period` returns `period` records.

Comment (marcjohnson-kint) [reply descriptors/ModuleDescriptor-template.json]: That is fair enough. I don't want to make this changer any bigger.

**Pull request merged**